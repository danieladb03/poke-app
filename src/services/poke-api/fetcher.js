export const baseUrl = "https://pokeapi.co/api/v2/";
const fetcher = (url) => fetch(`${baseUrl}${url}`).then((res) => res.json());

export default fetcher;
