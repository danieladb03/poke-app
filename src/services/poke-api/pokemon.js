import useSWR from 'swr';
import fetcher from './fetcher';

export const useFetchPokemons = () => {
  const {data, error} = useSWR('pokemon?limit=151', fetcher);

  // include query param con default vacio
  // useState para crear filteredPokemons
  // useEffect se va a llamar cuando el query cambie
  //    filtrar data, validar que data no sea null
  //    resultado setear a filteredPokemons
  // en lugar de devolver data devolver filteredPokemons

  return {
    pokemons: data?.results,
    isLoading: !error && !data,
    isError: error,
  };
};
