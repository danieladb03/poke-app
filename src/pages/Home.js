/* eslint-disable no-unused-vars */
import Container from "../core/Container";
import styled from "styled-components";
import Header from "../core/Header";
import PokeCard from "../core/PokeCard";
import Layout from "./Layout";
import {useFetchPokemons} from "../services/poke-api/pokemon";
import {useState} from "react";
import Loader from "../core/loader";

const pokeCards = [1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6];

const Wrapper = styled(Container)`
  width: calc(100% - 298px);
`;

const PokeCardsWrapper = styled(Container)`
  justify-content: "space-between";
  height: calc(100vh - 65px);
  padding: 55px 10px 0px 10px;
  overflow-y: auto;
`;

const Home = () => {
  const [query, setQuery] = useState("");
  const {pokemons, isLoading, isError} = useFetchPokemons(query);
  const handleChange = (event) => setQuery(event.target.value);

  return (
    <Layout>
      <Wrapper flexDirection="column">
        <Header>
          <Container justifyContent="flex-end" alignItems="center">
            <input
              type="text"
              name="query"
              placeholder="Pokedex search"
              onChange={handleChange}
            />
            <span className="material-icons md-light">search</span>
          </Container>
        </Header>
        <PokeCardsWrapper flexWrap="wrap">
          <Loader
            count={10}
            width="192px"
            height="200px"
            isLoading={isLoading}
            style={{margin: "6px"}}
          >
            {pokemons?.map((pokemon, k) => (
              <PokeCard key={k} pokemon={pokemon} />
            ))}
          </Loader>
        </PokeCardsWrapper>
      </Wrapper>
    </Layout>
  );
};

export default Home;
