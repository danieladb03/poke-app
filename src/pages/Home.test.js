// tipos de tests Snapshot y testing library
// npm test -- --coverage para ver el porcentaje de test de cada componente

import {fireEvent, render, screen} from "@testing-library/react";
import Home from "./Home";
import {useFetchPokemons} from "../services/poke-api/pokemon";
import renderer from "react-test-renderer";

jest.mock("../services/poke-api/pokemon");

describe("Home", () => {
  const pokemonsResp = {
    pokemons: [
      {name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/"},
      {name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/"},
      {name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/"},
      {name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/"},
    ],
    isLoading: false,
  };
  const filteredPokemonsResp = {
    pokemons: [
      {name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/"},
      {name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/"},
    ],
    isLoading: false,
  };

  test("returns pokemon cards", () => {
    useFetchPokemons.mockReturnValue(pokemonsResp);
    render(<Home />);
    const pokemons = screen.getAllByTestId("poke-card");

    expect(pokemons).toHaveLength(pokemonsResp.pokemons.length);
    expect(useFetchPokemons.mock.calls.length).toBe(1);
  });

  test("returns loading states when is loading", () => {
    const resp = {
      isLoading: true,
    };
    useFetchPokemons.mockReturnValue(resp);
    render(<Home />);
    const pokemons = screen.queryAllByTestId("poke-card");

    expect(pokemons).toHaveLength(0);
    expect(useFetchPokemons.mock.calls.length).toBe(1);
    const tree = renderer.create(<Home />).toJSON(); // Cómo debería verse el resultado final de la página
    expect(tree).toMatchSnapshot();
  });

  test("returns query search", () => {
    useFetchPokemons
      .mockReturnValueOnce(pokemonsResp)
      .mockReturnValue(filteredPokemonsResp);
    render(<Home />);
    const search = screen.getByRole("textbox");
    const query = "char";

    expect(search).toHaveAttribute("placeholder", "Pokedex search");
    expect(screen.queryAllByTestId("poke-card")).toHaveLength(4);

    fireEvent.change(search, {target: {value: query}});
    expect(screen.queryAllByTestId("poke-card")).toHaveLength(2);
    expect(useFetchPokemons).toHaveBeenCalledWith(query);
  });
});
