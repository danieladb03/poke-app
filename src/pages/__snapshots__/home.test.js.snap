// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home returns loading states when is loading 1`] = `
<div
  className="sc-bdfBwQ eXdLAz"
>
  <div
    className="sc-bdfBwQ sc-iBPRYJ ea-DoiS bAbfSZ"
  >
    <h1
      className="sc-dlfnbm gseDyA"
      color="#FFFFFF"
    >
      Poke
    </h1>
    <img
      className="sc-pFZIQ iFpWzT"
      src="profile.png"
    />
    <p
      className="sc-fubCfw drDFoG"
    >
      Daniela DB
    </p>
    <p
      className="sc-hKgILt icNjze"
      color="#FFFFFF"
    >
      Poke Master
    </p>
    <p
      className="sc-hKgILt ewhALV"
      color="#FFFFFF"
    >
      Lvl 25
    </p>
  </div>
  <div
    className="sc-bdfBwQ sc-jrAGrp hZVvYV hiYJFV"
  >
    <header
      className="sc-gsTCUz fNSlpe"
    >
      <div
        className="sc-bdfBwQ blFYpd"
      >
        <input
          name="query"
          onChange={[Function]}
          placeholder="Pokedex search"
          type="text"
        />
        <span
          className="material-icons md-light"
        >
          search
        </span>
      </div>
    </header>
    <div
      className="sc-bdfBwQ sc-kEjbxe exVEDH hojsNo"
    >
      <span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
        <span
          className="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          style={
            Object {
              "height": "200px",
              "margin": "6px",
              "width": "192px",
            }
          }
        >
          ‌
        </span>
      </span>
    </div>
  </div>
</div>
`;
